core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1364,11 +1364,19 @@ class ReplicaManager(val config: KafkaConfig,
                 Some(partition)
             }
 
-            // Next check partition's leader epoch
+            // Next check the topic ID and the partition's leader epoch
             partitionOpt.foreach { partition =>
               val currentLeaderEpoch = partition.getLeaderEpoch
               val requestLeaderEpoch = partitionState.leaderEpoch
-              if (requestLeaderEpoch > currentLeaderEpoch) {
+              val requestTopicId = topicIds.get(topicPartition.topic)
+              val (consistentTopicId, logTopicId) = partition.checkOrSetTopicId(requestTopicId)
```

Comment:
The return type here is a little awkward. It looks like we only need `logTopicId` for the log message below. Since `Partition` also has access to the state change logger, how about we move it into `checkOrSetTopicId`?

Reply:
Good point. I didn't realize it had access.